lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/JnaTrash.java

```diff
@@ -0,0 +1,34 @@
+package org.enso.desktopenvironment;
+
+import com.sun.jna.platform.FileUtils;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * JNA trash implementation calling native platform-dependent APIs.
+ *
+ * <p>Native {@link FileUtils} are implemented for Windows and macOS platforms.
+ */
+final class JnaTrash implements Trash {
+
+  @Override
+  public boolean isSupported() {
+    return FileUtils.getInstance().hasTrash();
```

Comment:
Mac OS X implementation is using:
- [CoreServices](https://github.com/java-native-access/jna/blob/a587921bac074b1bd1b0a0a5536587660a9b954e/contrib/platform/src/com/sun/jna/platform/mac/MacFileUtils.java#L45) library
- invokes [FSMoveObjectToTrashSync](https://github.com/java-native-access/jna/blob/a587921bac074b1bd1b0a0a5536587660a9b954e/contrib/platform/src/com/sun/jna/platform/mac/MacFileUtils.java#L82)

Reply:
Maven project capable to build `target/Trash` binary that moves provided file to Mac OS X trash:
```diff
diff --git a/pom.xml b/pom.xml
new file mode 100644
index 0000000..5da15f0
--- /dev/null
+++ b/pom.xml
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.apidesign.demo.trash</groupId>
+    <artifactId>Trash</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.compiler.source>21</maven.compiler.source>
+        <maven.compiler.target>21</maven.compiler.target>
+        <exec.mainClass>org.apidesign.demo.trash.Trash</exec.mainClass>
+    </properties>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.graalvm.buildtools</groupId>
+                <artifactId>native-maven-plugin</artifactId>
+                <version>0.10.2</version>
+                <extensions>true</extensions>
+                <executions>
+                    <execution>
+                        <id>build-native</id>
+                        <goals>
+                            <goal>compile-no-fork</goal>
+                        </goals>
+                        <phase>package</phase>
+                    </execution>
+                    <execution>
+                        <id>test-native</id>
+                        <goals>
+                            <goal>test</goal>
+                        </goals>
+                        <phase>test</phase>
+                    </execution>
+                </executions>
+                <configuration>
+                    <mainClass>org.apidesign.demo.trash.Trash</mainClass>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
\ No newline at end of file
diff --git a/src/main/java/org/apidesign/demo/trash/Trash.java b/src/main/java/org/apidesign/demo/trash/Trash.java
new file mode 100644
index 0000000..e5edc7c
--- /dev/null
+++ b/src/main/java/org/apidesign/demo/trash/Trash.java
@@ -0,0 +1,55 @@
+package org.apidesign.demo.trash;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.graalvm.nativeimage.UnmanagedMemory;
+import org.graalvm.nativeimage.c.CContext;
+import org.graalvm.nativeimage.c.function.CFunction;
+import org.graalvm.nativeimage.c.function.CLibrary;
+import org.graalvm.nativeimage.c.type.CCharPointer;
+import org.graalvm.nativeimage.c.type.CTypeConversion;
+import org.graalvm.word.Pointer;
+import org.graalvm.word.WordFactory;
+
+// @CContext(Trash.TrashDirectives.class)
+// @CContext(PosixDirectives.class)
+@CLibrary("-framework CoreServices")
+public class Trash {
+
+    public static void main(String[] args) {
+        System.out.println("Moving " + args[0] + " to trash");
+        Pointer ref = WordFactory.nullPointer();
+        Pointer source = UnmanagedMemory.malloc(80);
+        Pointer target = UnmanagedMemory.malloc(80);
+
+        var kFSPathMakeRefDoNotFollowLeafSymlink = 0x01;
+        var kFSFileOperationDefaultOptions = 0x00;
+        try (var path = CTypeConversion.toCString(args[0])) {
+            var r1 = FSPathMakeRefWithOptions(path.get(), kFSPathMakeRefDoNotFollowLeafSymlink, source, ref);
+            System.err.println("result one: " + r1);
+            var r2 = FSMoveObjectToTrashSync(source, target, kFSFileOperationDefaultOptions);
+            System.err.println("result two: " + r2);
+        }
+    }
+
+    @CFunction
+    static native int FSPathMakeRefWithOptions(CCharPointer path, int flags, Pointer buffer, Pointer any);
+
+    @CFunction
+    static native int FSMoveObjectToTrashSync(Pointer source, Pointer target, int flags);
+
+    public static final class TrashDirectives implements CContext.Directives {
+
+        @Override
+        public List<String> getHeaderFiles() {
+            return Arrays.asList("</Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/CarbonCore.framework/Versions/A/Headers/Files.h>");
+        }
+
+        @Override
+        public List<String> getLibraries() {
+            return Arrays.asList("CoreServices");
+        }
+
+    }
+}
```

Reply:
Improved version that also **conditionally enables itself** only on Mac:
```java
package org.apidesign.demo.trash;

import java.util.Arrays;
import java.util.List;

import org.graalvm.nativeimage.UnmanagedMemory;
import org.graalvm.nativeimage.c.CContext;
import org.graalvm.nativeimage.c.constant.CConstant;
import org.graalvm.nativeimage.c.function.CFunction;
import org.graalvm.nativeimage.c.type.CCharPointer;
import org.graalvm.nativeimage.c.type.CTypeConversion;
import org.graalvm.word.Pointer;
import org.graalvm.word.WordFactory;

@CContext(Trash.CoreServices.class)
public class Trash {

    public static void main(String[] args) {
        System.out.println("Moving " + args[0] + " to trash");
        Pointer ref = WordFactory.nullPointer();
        Pointer source = UnmanagedMemory.malloc(80);
        Pointer target = UnmanagedMemory.malloc(80);

        try (var path = CTypeConversion.toCString(args[0])) {
            var r1 = FSPathMakeRefWithOptions(path.get(), kFSPathMakeRefDoNotFollowLeafSymlink(), source, ref);
            System.err.println("result one: " + r1);
            var r2 = FSMoveObjectToTrashSync(source, target, kFSFileOperationDefaultOptions());
            System.err.println("result two: " + r2);
        }
        UnmanagedMemory.free(source);
        UnmanagedMemory.free(target);
    }

    @CConstant
    static native int kFSPathMakeRefDoNotFollowLeafSymlink();

    @CConstant
    static native int kFSFileOperationDefaultOptions();

    @CFunction
    static native int FSPathMakeRefWithOptions(CCharPointer path, int flags, Pointer buffer, Pointer any);

    @CFunction
    static native int FSMoveObjectToTrashSync(Pointer source, Pointer target, int flags);

    public static final class CoreServices implements CContext.Directives {
        @Override
        public boolean isInConfiguration() {
            return System.getProperty("os.name").contains("Mac");
        }

        @Override
        public List<String> getHeaderFiles() {
            return Arrays.asList("<CoreServices/CoreServices.h>");
        }

        @Override
        public List<String> getLibraries() {
            return Arrays.asList("-framework CoreServices");
        }

    }
}
```